src/orange/orangeinp/detail/BoundingZone.hh

```diff
@@ -0,0 +1,113 @@
+ *
+ * This class partition space into "known inside" a volume, "maybe inside,
+ * maybe outside", and "known outside".
+ *
+ * - Known outside is the typical bounding box case: nowhere outside the
+ * bounding box is inside the volume; i.e., the box completely encloses the
+ * volume (but may also enclose parts that *aren't* volume).
+ * - Known inside can be used for safety distance calculation.
+ * - Flipping the two cases when applying a "not" operator to a CSG node: so
+ *   "known inside" becomes "known outside" and could potentially be flipped
+ *   back in a subsequent simplification.
+ *
+ * The \c exterior box always encloses (or is identical to) \c interior in the
+ * bounding zone. If \c negated is \c false, then the exterior box is what we
+ * typically think of as a bounding box. If \c true, then even though the
+ * exterior bounding box is larger than the interior box, *all points inside
+ * the interior box are known to be outside the volume*.  Regardless of
+ * negation, the area between the interior and exterior bounding boxes is \em
+ * indeterminate: a point in that space may be within the CSG region, or it may
+ * not.
+ *
+ * The behavior of boundaries (whether coincident boundary/shape should count
+ * as being "interior" or "exterior") shouldn't matter because we will always
+ * "bump" boundaries before using them for transport.
+ *
+ * The following table shows the semantics of set operations on the bounding
+ * zones.  Below, \c ~zone means a zone with the \c negated flag set; \c ~box
+ * means "complement of a bbox": all space *except*
+ * inside box; and the union operator implies the (possibly disconnected and
+ * probably not box-shaped!) union of the two boxes.
+ *
+ * | Zone      | known inside   | known outside   |
+ * | ------    | ------------   | -------------   |
+ * | `A`       | `A_i`          | `~A_x`          |
+ * | `~A`      | `~A_x`         | `A_i`           |
+ * | `A & B`   | `A_i & B_i`    | `~(A_x & B_x)`  |
+ * | `~A & B ` | `~A_x & B_i`   | `~(~A_i & B_x)` |
+ * | `~A & ~B` | `~A_x & ~B_x`  | `~(~A_i & ~B_i)`|
+ * | `A | B`   | `A_i | B_i`    | `~(A_x | B_x)`  |
+ * | `~A | B ` | `~A_x | B_i`   | `~(~A_i | B_x)` |
+ * | `~A | ~B` | `~A_x | ~B_x`  | `~(~A_i | ~B_i)`|
+ *
+ * The following set algebra can be used:
+ * - Involution: \verbatim ~~A <=> A \endverbatim
+ * - De Morgan's law 1: \verbatim ~A | ~B <=> ~(A & B) \endverbatim
+ * - De Morgan's law 1:\verbatim ~A & ~B <=> ~(A | B) \endverbatim
+ * - Set difference: \verbatim A & ~B <=> A - B \endverbatim
+ * - Negated set difference: \verbatim A | ~B <=> ~(B - A) \endverbatim
+ *
+ * When \c negated is set, the *inside* of the interior bounding box is "always
+ * outside" and the *outside* of the exterior is "always inside".
+ *
+ * The default bounding zone places all points in the \em indeterminate zone:
+ * the exterior is the "universe set" (infinite) and the interior is the "empty
+ * set" (null).
+ */
+struct BoundingZone
```

Comment:
I was thinking it's a zone that partitions "inside" from "outside", rather than a *box* that partitions inside from outside. Wouldn't partitions be potentially confusing with BIH/BVH?

Reply:
Okay, so the "zone" is technically region between the two bounding boxes then? I like that a lot actually --- I think you should state that explicitly in the doc string.